src/mapping/config/MappingConfiguration.cpp

```diff
@@ -24,134 +25,169 @@
+inline constexpr bool always_false_v = false;
+
+// Reads and returns an attribute value of any type (bool, string, double) from the tag,
+// if the attribute is actually contained in the tag. If the attribute is not in the tag,
+// an \p initializationValue is returned.
+template <typename AttributeType>
+AttributeType getAttributeIfPresent(xml::XMLTag &tag, const std::string &attributeName, AttributeType initializationValue)
+{
+  AttributeType value = initializationValue;
+  if (tag.hasAttribute(attributeName)) {
+    using T = std::decay_t<AttributeType>;
+    if constexpr (std::is_same_v<T, double>) {
+      value = tag.getDoubleAttributeValue(attributeName);
+    } else if constexpr (std::is_same_v<T, bool>) {
+      value = tag.getBooleanAttributeValue(attributeName);
+    } else if constexpr (std::is_same_v<T, std::string>) {
+      value = tag.getStringAttributeValue(attributeName);
+    } else {
+      static_assert(always_false_v<T>, "Attribute type is not supported.");
+    }
   }
-  {
-    XMLTag tag(*this, VALUE_RBF_GAUSSIAN, occ, TAG);
-    tag.setDocumentation("Local radial-basis-function mapping based on the Gaussian RBF using a cut-off threshold.");
-    tag.addAttribute(makeXMLAttribute<double>(ATTR_SHAPE_PARAM, std::numeric_limits<double>::quiet_NaN())
-                         .setDocumentation("Specific shape parameter for RBF basis function."));
-    tag.addAttribute(makeXMLAttribute(ATTR_SUPPORT_RADIUS, std::numeric_limits<double>::quiet_NaN())
-                         .setDocumentation("Support radius of each RBF basis function (global choice)."));
-    tags.push_back(tag);
+  return value;
+}
+
+// Enum required for the RBF instantiations
+enum struct RBFBackend {
+  Eigen,
+  PETSc
+};
+
+// Helper in order to resolve the template instantiations.
+// Only the template specializations are of interest
+template <RBFBackend Backend, typename RBF>
+struct BackendSelector {
+  typedef RBF type;
+};
+
+// Specialization for the RBF Eigen backend
+template <typename RBF>
+struct BackendSelector<RBFBackend::Eigen, RBF> {
+  typedef mapping::RadialBasisFctMapping<RBF> type;
+};
+
+// Specialization for the PETSc RBF backend
+#ifndef PRECICE_NO_PETSC
+template <typename RBF>
+struct BackendSelector<RBFBackend::PETSc, RBF> {
+  typedef mapping::PetRadialBasisFctMapping<RBF> type;
+};
+#endif
+
+// The actual instantion helper, which avoids enumerating all mapping implementations (more will come) with all RBF kernels
+// The first three arguments of the constructor are prescribed: constraint, dimension and the RBF function object, all other
+// constructor arguments are just forwareded. The first argument (BasisFunction) indicated then the actual instantiation to return.
+template <RBFBackend T, typename... Args>
+PtrMapping instantiateRBFMapping(BasisFunctions functionType, mapping::Mapping::Constraint &constraint, int dimension, double supportRadius, double shapeParameter,
+                                 Args &&... args)
```

Comment:
I'm not completely sure I understand what you mean here. Are you suggesting to do a type overloading with `std::variant` for all  RBF functions? I could imagine that things get messy. Since the function type is a template parameter of the rbf implementation, we have to instantiate, i.e., put all functions into the control flow at some point, right?

Reply:
OK, let me clarify and fix a bug in my example above:

```cpp
template <RBFBackend T, typename RBFunc, typename... Args>
PtrMapping instantiateRBFMapping(
  RBFunc func, // this could be mapping::CompactPolynomialC0
  mapping::Mapping::Constraint &constraint,
  int dimension,
  Args &&... args)
{
  using Backend = typename BackendSelector<T, RBFunc>::type; // makes things easier to follow
  return PtrMapping(new Backend(constraint, dimension, std::move(func), std::forward<Args>(args)...));
}
```

Then the variant can be used as such
```cpp
using AnyRBFFunction = std::variant<Multiquadrics, InverseMultiquadrics , CompactPolynomialC0 /*, ...*/ >;

// you can handle erros in there too
template <typename RBF>
AnyRBFFunction constructRBF(double shapeparameter, double supportradius);

// to use
auto func = constructRBF(...);
auto mappingPtr = std::visit([&](const auto& func) {
return instantiateRBFMapping(
  func,
  constraint,
  dimension,
  args...); // not sure how to pass the pack here
});
```

I hope that clarifies what I am aiming for :sweat_smile:

Reply:
Are you sure this will work out? Note that in your example, `auto func` will still be of type `variant`, so we cannot use it with `instantiateRBFMapping`, but would need to use `std::get<T>` beforehand (which would elide the actual advantage again. If we make the function `instantiateRBFMapping` accept the variant as a type, I'm pretty sure the template argument deduction of the BackendSelector will fail to deduce the RBF function type. Am I missing something?

Reply:
Ok, makes more sense now. However, then I end up having the switch statement inside the `constructRBF` function, right?

Reply:
Yep. It will be somewhere, but at least there is as little logic inside as possible.

Reply:
In the end I decided against this. You said

> Then the entire switch is contained in a single function

which is also the case right now?

and 

> Just a few comments on reducing the overall size of the code. 

Using your suggestion will add additional spaghetti code in order to forward the basis function into the variant and finally into  the visitor (it will definitely be more code).

and 

>  but at least there is as little logic inside as possible.

I don't see how the additional logic helps here. IMHO it just introduced more complexity when trying to follow the code path, hence, and a code which is harder to read.

Reply:
> which is also the case right now?

Yes, but it contains additional logic.

> Using your suggestion will add additional spaghetti code in order to forward the basis function into the variant and finally into the visitor (it will definitely be more code).

The point is to split the logic for parsing, handling arguments and building mappings.
This way we end up with little unit-testable fragments. In a way, it's the opposite of Spaghetti code.

Dispatching to the visitor looks daunting, but it is just forwarding things, and the code doesn't need to be touched when adding new BasisFunctions.

> IMHO it just introduced more complexity when trying to follow the code path, hence, and a code which is harder to read.

* extract rbf name and settings
* use them to build a rbf function
* use it to instantiate an RBF mapping
* return the mapping

Logical data flow and function that have their clear sets of inputs and outputs.
Isn't that easier to read?

Reply:
> and the code doesn't need to be touched when adding new BasisFunctions.

Then we have to change things in two places instead of one, because we have to touch the visitor type, and the switch.

>  Isn't that easier to read?

No. We have

>    extract rbf name and settings
    use them to build a rbf function
    use it to instantiate an RBF mapping
    return the mapping

vs (the version now)

 > extract rbf name and settings
    use instantiate an RBF mapping with kernel function
    return the mapping

Anyhow, I don't want to discuss it any further. I will change it.

Reply:
> Then we have to change things in two places instead of one, because we have to touch the visitor type, and the switch.

You mean the variant type. True. It's less logic though and the compiler will complain about the missing variant.

> use instantiate an RBF mapping with kernel function

Splitting this into two steps, makes them easier to test though. Especially once we start testing error conditions.

Reply:
Done